spec_summary/schema.go

```diff
@@ -1,17 +1,201 @@
+	DirectedFilters DirectedFilterCounts `json:"directed_filters"`
+
+	// Filters that are independent of the request/response.
+	NondirectedFilters NondirectedFilterCounts `json:"nondirected_filters"`
+}
+
+func NewSummaryByDirection() *SummaryByDirection {
+	return &SummaryByDirection{
+		DirectedFilters: DirectedFilterCounts{
+			RequestDirection:  make(NondirectedFilterCounts),
+			ResponseDirection: make(NondirectedFilterCounts),
+		},
+		NondirectedFilters: make(map[FilterKind]map[FilterValue]int),
+	}
+}
+
+func (s *SummaryByDirection) ToSummary() *Summary {
+	if s == nil {
+		return nil
+	}
+
+	return &Summary{
+		// Non-directional properties.
+		Directions:  s.NondirectedFilters[DirectionFilter],
+		Hosts:       s.NondirectedFilters[HostFilter],
+		HTTPMethods: s.NondirectedFilters[HttpMethodFilter],
+		Paths:       s.NondirectedFilters[PathFilter],
+
+		// Directional properties.
+		Authentications: s.DirectedFilters.GetCountsByValue(RequestDirection, AuthFilter),
+		DataKinds:       s.DirectedFilters.MergeAcrossDirections(DataFormatKindFilter),
+		DataFormats:     s.DirectedFilters.MergeAcrossDirections(DataFormatFilter),
+		DataTypes:       s.DirectedFilters.MergeAcrossDirections(DataTypeFilter),
+		Params:          s.DirectedFilters.MergeAcrossDirections(ParamFilter),
+		Properties:      s.DirectedFilters.MergeAcrossDirections(PropertyFilter),
+		ResponseCodes:   s.DirectedFilters.GetCountsByValue(ResponseDirection, ResponseCodeFilter),
+	}
+}
+
+type NondirectedFilterCounts map[FilterKind]map[FilterValue]int
+
+func (cs NondirectedFilterCounts) Increment(kind FilterKind, v FilterValue) {
+	byVal, ok := cs[kind]
+	if !ok {
+		byVal = make(map[FilterValue]int)
+		cs[kind] = byVal
+	}
+
+	byVal[v] += 1
+}
+
+func (cs NondirectedFilterCounts) Insert(kind FilterKind, v FilterValue, count int) {
+	byVal, ok := cs[kind]
+	if !ok {
+		byVal = make(map[FilterValue]int)
+		cs[kind] = byVal
+	}
+
+	byVal[v] = count
+}
+
+// Calls f on each kind, value.  Returns false immediately if f returns false
+// or true otherwise.
+func (cs NondirectedFilterCounts) ForEach(f func(kind FilterKind, v FilterValue, count int) bool) bool {
+	for kind, byVal := range cs {
+		for v, count := range byVal {
+			if !f(kind, v, count) {
+				return false
+			}
+		}
+	}
+
+	return true
+}
+
+type DirectedFilterCounts map[Direction]NondirectedFilterCounts
+
+func (cs DirectedFilterCounts) Increment(direction Direction, kind FilterKind, v FilterValue) {
+	byKind, ok := cs[direction]
+	if !ok {
+		byKind = make(NondirectedFilterCounts)
+		cs[direction] = byKind
+	}
+
+	byKind.Increment(kind, v)
+}
+
+func (cs DirectedFilterCounts) Insert(direction Direction, kind FilterKind, v FilterValue, count int) {
+	byKind, ok := cs[direction]
+	if !ok {
+		byKind = make(NondirectedFilterCounts)
+		cs[direction] = byKind
+	}
+
+	byKind.Insert(kind, v, count)
+}
+
+// Calls f on each direction, kind, value.  Returns false immediately if f returns false
+// or true otherwise.
```

Comment:
```suggestion
// Calls f on each (direction, kind, value) tuple. If f returns false, iteration
// stops immediately, and false is returned. Otherwise, returns true.
```